contracts/BC_fusion/AirDrop.sol

```diff
@@ -0,0 +1,146 @@
+pragma solidity 0.8.17;
+
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+import "./interface/ITokenHub.sol";
+import "./interface/IAirDrop.sol";
+import "./System.sol";
+import "./lib/Utils.sol";
+
+contract AirDrop is IAirDrop, ReentrancyGuardUpgradeable, System {
+    string public constant sourceChainID = "Binance-Chain-Ganges";
+    address public approvalAddress = 0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa;
+    bytes32 public override merkleRoot = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    bool public merkleRootAlreadyInit = false;
+
+    // claimedMap is used to record the claimed token.
+    mapping(bytes32 => bool) private claimedMap;
+
+    function isClaimed(bytes32 node) public view override returns (bool) {
+        return claimedMap[node];
+    }
+
+    function claim(
+        bytes32 tokenSymbol, uint256 amount,
+        bytes calldata ownerPubKey, bytes calldata ownerSignature, bytes calldata approvalSignature,
+        bytes32[] calldata merkleProof) nonReentrant external override {
+        // Recover the owner address and check signature.
+        bytes memory ownerAddr = _verifyTMSignature(ownerPubKey, ownerSignature, _tmSignatureHash(tokenSymbol, amount, msg.sender));
+        // Generate the leaf node of merkle tree.
+        bytes32 node = keccak256(abi.encodePacked(ownerAddr, tokenSymbol, amount));
+    
+        // Check if the token is claimed.
+        require(isClaimed(node), "AlreadyClaimed");
+
+        address contractAddr = address(0x00);
+        // Check if the token is exist.
+        if (tokenSymbol != ITokenHub(TOKEN_HUB_ADDR).BEP2_TOKEN_SYMBOL_FOR_BNB()) {
+            contractAddr = _checkTokenContractExist(tokenSymbol);
+        }
+        
+        // Verify the approval signature.
+        _verifySignature(msg.sender, ownerSignature, approvalSignature, node, merkleProof);
+    
+        // Verify the merkle proof.
+        require(MerkleProof.verify(merkleProof, merkleRoot, node), "InvalidProof");
+    
+        // Mark it claimed and send the token.
+        claimedMap[node] = true;
+        
+        // Unlock the token from TokenHub.
+        ITokenHub(TOKEN_HUB_ADDR).unlock(contractAddr, msg.sender, amount);
+
+        emit Claimed(tokenSymbol, msg.sender, amount);
+    }
+
+    function _verifySignature(address account, bytes memory ownerSignature, bytes memory approvalSignature, bytes32 leafHash, bytes32[] memory merkleProof) private view {
+        // Ensure the account is not the zero address
+        require(account != address(0), "InvalidAddress");
```

Comment:
I think such check is useless, because it have to be msg.sender, it must be a vaild EOA or contract

Reply:
fixed